frontend/src/useFlow.ts

```diff
@@ -2,45 +2,65 @@ import { useCallback, useState } from 'react';
-import { buildSegment, extractParagraphs, postSegments } from './services/segment.service';
-import { getSource, postSource } from './services/source.service';
+import { buildSegment, extractParagraphs, extractText, ExtractTextResult, postSegments } from './services/segment.service';
+import { getSource, postSource, postSourceOriginLinks } from './services/source.service';
 import { Segment, Source } from './types/frontend-types';
 
+export interface AdditionalSourceInfo {
+  file: File;
+  name: string;
+  language: string;
+  id?: number;
+}
+
 // import compareTwoStrings from 'string-similarity-js';
 const DEFAULT_PROMPT_KEY = 'prompt_1';
+const ADDITIONAL_SOURCES_TEXT_LENGTH_MULTIPLIER = 1.5;
 
 const buildAndSaveSegments = async (
   paragraphs: string[],
   source_id: number,
   properties: Record<string, any>,
-  originalSegments?: Segment[]
+  originalSegments?: Segment[],
+  additional_sources_segments?: Record<string, string[]>
 ): Promise<Segment[]> => {
-  const segments = paragraphs.map((text, index) => {
-    const originalSegment = originalSegments?.[index];
-    return buildSegment({
-      text,
-      source_id,
-      order: originalSegment?.order ?? index + 1,
-      properties,
-      original_segment_id: originalSegment?.id,
-      original_segment_timestamp: originalSegment?.timestamp
-    });
+  return await postSegments({
+    paragraphs,
+    source_id,
+    properties,
+    originalSegments,
+    additional_sources_segments
   });
-  return await postSegments(segments);
 };
 
-const initSourceFromFile = async (file: File, name: string, originalLanguage: string, translatedLanguage: string, dictionaryId: null|number, dictionaryTimestamp: null|string)
-  : Promise<{originalSegments: Segment[], translatedSourceId: number}> => {
+const initSourceFromFile = async (file: File, name: string, originalLanguage: string, translatedLanguage: string, dictionaryId: null|number, dictionaryTimestamp: null|string, additionalSources: AdditionalSourceInfo[] = [])
+  : Promise<{originalSegments: Segment[], translatedSourceId: number, additionalSourcesText: ExtractTextResult[]}> => {
   // Create original and translation sources
-  const { originalSourceId, translatedSourceId } = await createSources(name, originalLanguage, translatedLanguage, dictionaryId, dictionaryTimestamp);
+  const { originalSourceId, translatedSourceId, additionalSourceIds } = await createSources(name, originalLanguage, translatedLanguage, dictionaryId, dictionaryTimestamp, additionalSources);
 
   // Extract paragraphs from uploaded file
   const { paragraphs, properties } = await extractParagraphs(file);
 
+  // Extract text from all additional sources and store temporal segments
+  let additionalSourcesText: ExtractTextResult[] = [];
+  if (additionalSources.length > 0) {
+    additionalSources.forEach((source, index) => {
+      source.id = additionalSourceIds[index];
+    });
+    const extractedTexts = await extractText(additionalSources);
```

Comment:
Why not use `extractParagraphs` several times? This is much simpler and no need for so many redundant code.